File: 2021/haskell/day8.hs
{-# LANGUAGE TupleSections, BangPatterns, FlexibleContexts #-}
import Control.Arrow
import Data.Functor
import Control.Monad.State
import Data.List
import qualified Data.Set as S
import Control.Monad
import Data.Maybe

sample :: String
sample = "be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe\n\
\edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc\n\
\fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg\n\
\fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb\n\
\aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea\n\
\fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb\n\
\dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe\n\
\bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef\n\
\egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb\n\
\gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce"

parseLine :: String -> ([String], [String])
parseLine =  break (=='|') >>> second tail >>> both (groupBy (\a b -> a /= ' ' && b /= ' ') >>> filter (/=" "))

both :: Arrow a => a b c -> a (b, b) (c, c)
both = join (***)

partOne :: [([String], [String])] -> Int
-- 1 takes 2, 7 takes 3, 4 takes 4, 8 takes 7
partOne = map  (snd >>> map length >>> filter (`elem` [2, 3, 4, 7]) >>> length) >>> sum

parse :: String -> [([String], [String])]
parse = lines >>> map parseLine



fLength :: Int -> [[a]] -> Maybe [a]
fLength x = find (length >>> (==x))

fRelative :: String -> Int -> [String] -> Maybe String
fRelative a i = find ((\\ a) >>> length >>> (== i))


-- get 1, 4, 7, 8 from their unique counts
stage1 :: MonadState [String] m => m (String, String, String, String)
stage1 = do
  xs <- get
  let Just !one = fLength 2 xs
      Just !four = fLength 4 xs
      Just !seven = fLength 3 xs
      Just !eight = fLength 7 xs
  modify' $ filter (\x -> x /= one && x /= four && x /= seven && x /= eight)
  return (one, four, seven, eight)

stage2 :: MonadState [String] m => String -> m (String, String)
stage2 seven = do
  xs <- get
  let Just !three = fRelative seven 2 xs
      Just !six   = fRelative seven 4 xs
  modify' $ filter $ \x -> x /= three && x /= six
  return (three, six)

-- Get two and zero from 3 and 4
stage3 :: MonadState [String] m => (String, String) -> m (String, String, String, String)
stage3 (three, four) = do
  (xs, rest) <- gets (partition ((\\ four) >>> length >>> (==3)))
  let [a, b] = xs
  let [c, d] = rest
  let lenA = length $ a \\ three
      lenC = length c
      (two, zero) = if lenA == 2 then (b, a) else (a, b)
      (five, nine) = if lenC == 6 then (d, c) else (c, d)

  return (two, zero, five, nine)


mkTable :: [String] -> [(S.Set Char, Int)]
mkTable = evalState $ do
  (one, four, seven, eight) <- stage1
  (three, six) <- stage2 seven
  (two, zero, five, nine) <- stage3 (three, four)
  return $ [ (zero, 0), (one, 1), (two, 2), (three, 3), (four, 4), (five, 5), (six, 6), (seven, 7), (eight, 8), (nine, 9) ] <&> first S.fromList


          
getValue :: [(S.Set Char, Int)] -> String -> Int
getValue table = S.fromList >>> flip lookup table >>> fromJust


decode :: [(S.Set Char, Int)] -> [String] -> Int
decode table = map (getValue table) >>> foldl (\x i -> x * 10 + i) 0

decodeLine :: ([String], [String]) -> Int
decodeLine (samples, code) = 
  let !table = mkTable samples
  in decode table code


partTwo :: [([String], [String])] -> Int
partTwo = map decodeLine >>> sum


main :: IO ()
main = do
  contents <- readFile "inputs/day8.txt" <&> parse
  putStrLn "Part 1:"
  print $ partOne contents
  putStrLn "Part 2:"
  print $ partTwo contents

File: 2021/python/day8.py
def get_digit(one: set[int], four: set[int], unknown: set[int]) -> int:
    match len(unknown):
        case 2:
            return 1
        case 3:
            return 7
        case 4:
            return 4
        case 7:
            return 8
        case 5:  # candidates: 2, 3, 5
            # 3 is the only one that has both segments in common with 1
            if len(unknown.intersection(one)) == 2:
                return 3
            # 2 is the only one that has two segments in common with 4
            elif len(unknown.intersection(four)) == 2:
                return 2
            # 5 coincides with 3 in 1 and with 2 on 4 so left to elimination
            else:
                return 5

        case 6:  # candidates: 6, 9, 0
            # 6 is the only one that has just 1 segment in common with 1; rest has 2
            if len(unknown.intersection(one)) == 1:
                return 6
            # 9 is the only one that contains four, the rest has 3 in common with it.
            elif len(unknown.intersection(four)) == 4:
                return 9
            # 0 is like 5, it is not "the only one" in any of the upper cases,  so it is left to elimination
            else:
                return 0


def decode(one: set[int], four: set[int], target: list[set[int]]) -> int:
    result = 0
    for unknown in target:
        digit = get_digit(one, four, unknown)
        result = result * 10 + digit

    return result


def decode_line(codes: list[set[int]], target: list[set[int]]) -> int:
    one, four = find_one_and_four(codes)
    return decode(one, four, target)


def parse_line(l: str):
    l, _, r = l.partition(" | ")
    return [*map(set, l.split())], [*map(set, r.split())]


def find_one_and_four(l: list[set[int]]) -> tuple[set[int], set[int]]:
    one = [*filter(lambda x: len(x) == 2, l)][0]
    four = [*filter(lambda x: len(x) == 4, l)][0]
    return one, four


with open("inputs/day8.txt") as file:
    lines = [*map(parse_line, file)]

    part_one_count = 0
    part_two_sum = 0
    for codes, target in lines:
        for value in target:
            if len(value) in [2, 3, 4, 7]:
                part_one_count += 1

        part_two_sum += decode_line(codes, target)

    print(f"Part 1: {part_one_count}")
    print(f"Part 2: {part_two_sum}")
